x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/RoleMappingMetadata.java

```diff
@@ -16,22 +16,46 @@
 import org.elasticsearch.common.collect.Iterators;
 import org.elasticsearch.common.io.stream.StreamInput;
 import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.common.xcontent.ChunkedToXContent;
 import org.elasticsearch.common.xcontent.ChunkedToXContentHelper;
+import org.elasticsearch.xcontent.ConstructingObjectParser;
+import org.elasticsearch.xcontent.ParseField;
 import org.elasticsearch.xcontent.ToXContent;
+import org.elasticsearch.xcontent.XContentBuilder;
+import org.elasticsearch.xcontent.XContentParser;
 import org.elasticsearch.xpack.core.security.authc.support.mapper.ExpressionRoleMapping;
 
 import java.io.IOException;
 import java.util.EnumSet;
 import java.util.Iterator;
+import java.util.LinkedHashSet;
+import java.util.List;
 import java.util.Objects;
 import java.util.Set;
 
 import static org.elasticsearch.cluster.metadata.Metadata.ALL_CONTEXTS;
+import static org.elasticsearch.xcontent.ConstructingObjectParser.constructorArg;
 
-public final class RoleMappingMetadata extends AbstractNamedDiffable<Metadata.Custom> implements Metadata.Custom {
+public final class RoleMappingMetadata extends AbstractNamedDiffable<Metadata.Custom> implements Metadata.Custom, ToXContent {
 
     public static final String TYPE = "role_mappings";
 
+    @SuppressWarnings("unchecked")
+    private static final ConstructingObjectParser<RoleMappingMetadata, Void> PARSER = new ConstructingObjectParser<>(
+        TYPE,
+        // serialization tests rely on the order of the ExpressionRoleMapping
+        args -> new RoleMappingMetadata(new LinkedHashSet<>((List<ExpressionRoleMapping>) args[0]))
+    );
+
+    static {
+        PARSER.declareObjectArray(
+            constructorArg(),
+            // role mapping names are lost when the role mapping metadata is serialized
+            (p, c) -> ExpressionRoleMapping.parse("name_not_available_after_deserialization", p),
```

Comment:
This further tightly couples the payload of our REST API to the payload of cluster state, despite the workflow being disjoint. I think this is fine for now, but it does present some long term concern w.r.t. these objects and the tight coupling of their serialization. The mismatch between persisted state across multiple versions of ES vs. transient state in a REST API yields different expectations and want to protect against good nature changes in the REST API breaking cluster state and vice-versa. Given the impact of getting this wrong at some future point, can you follow up with something that decouples these two ? Perhaps an adaptor pattern, where today it is an identity adaption (e.g. no differences) but is explicit and tests that changes to one will not influence the other.

Reply:
You're right to point out that this is an issue. I'll follow-up with an implementation of "`ExpressionRoleMapping` in the CS" separate from "`ExpressionRoleMapping` in request payload".